### PR TITLE
Delay imports in api modules if toolkit will be imported

### PR DESCRIPTION
**Pull request by corranwebster**
This is uses module `__getattr__` and `__dir__` based on [PEP 562](https://peps.python.org/pep-0562/) which is available now we are dropping Python 3.6 support.

Basic strategy is to handle two classes of imports - those which come directly via the toolkit object, and those which are via an import relative to the api module.  This allows us to eventually deprecate and remove stub modules which do nothing more than import using the toolkit, but we don't do that in this PR.

We test by checking that every public attribute returned by `dir()` can be accessed.

Includes a few drive-by fixes to avoid toolkit selection where where possible.  The most significant is adding formal `Interface` classes for `ActionManager` and its subclasses (and using those on `ApplicationWindow`).  It also cleans up a deprecated trait on `ApplicationWindow`.

### CODE REVIEW

**Comment by mdickinson** [comment]
One thought on the general idea: I think I've occasionally used an `import pyface.api` in the past to deliberately _force_ toolkit selection to occur. (Usually to avoid weird interactions when the toolkit selection happens at a time out of my control.) With this PR, what would be the recommended way to explicitly force toolkit selection?

File: pyface/tasks/task_window.py [hunk 1]
```diff
@@ -454,9 +457,9 @@ class TaskState(HasStrictTraits):
 
     central_pane = Instance(ITaskPane)
     dock_panes = List(IDockPane)
-    menu_bar_manager = Instance(MenuBarManager)
-    status_bar_manager = Instance(StatusBarManager)
-    tool_bar_managers = List(ToolBarManager)
+    menu_bar_manager = Instance(IMenuBarManager)
+    status_bar_manager = Instance(IStatusBarManager)
+    tool_bar_managers = List(IToolBarManager)
```
Comment (mdickinson) [inline pyface/tasks/task_window.py:462]: Minor drive-by suggestion
```suggestion
    tool_bar_managers = List(Instance(IToolBarManager))
```

**Review by mdickinson — APPROVED** [review]
LGTM; a few nitpick-level comments

The general approach looks good to me. We should be on the lookout for possible unintended downstream consequences arising from imports-with-side-effects now possibly happening at a different time from before. I don't have anything specific to point to, but we've seen weird interactions before from things like changing the import time of std. lib. modules like `concurrent.futures` and `logging`, or anything that registers an `atexit` handler at import time, or things that access `sys.stdout` (e.g., as a default parameter before vs after IPython has switched it out for something else).

Nothing actionable here, except possibly a sentence in the release notes warning people to look out for such problems.

File: pyface/action/action_manager.py [hunk 2]
```diff
@@ -45,7 +45,7 @@ class ActionManager(HasTraits):
     enabled = Bool(True)
 
     #: All of the contribution groups in the manager.
-    groups = Property(List(Group))
+    groups = Property(List(Instance(Group)), observe='_groups.items')
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/action/action_manager.py:48]: Not entirely unrelated - this PR introduces an `IActionManager` interface and this brings the `ActionManger` into agreement with that.  The `observe` is sort of orthogonal, but the advertised interface is `List(Instance(Group))` so it would be potentially surprising if the property didn't fire trait change events.

File: pyface/action/i_action_manager.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    # Events ----
+
+    #: fixme: We probably need more granular events than this!
+    changed = Event()
+
+    # ------------------------------------------------------------------------
+    # 'object' interface.
+    # ------------------------------------------------------------------------
+
+    def __init__(self, *args, **traits):
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/action/i_action_manager.py:63]: I had thought about that, but the style of `Manager(item0, item1, item2, ...)` seems to be part of the interface, for better or worse.  A toolkit which didn't support this in their version of an action manager would not be compatible with a lot of use-cases.

**Comment by corranwebster** [comment]
> With this PR, what would be the recommended way to explicitly force toolkit selection?

I've added `find_toolkit` to `api.py` so I think a reasonable way of doing this would be
```
from pyface.api import find_toolkit
find_toolkit()
```

This is idempotent, so the same toolkit object will be found no matter what.
